### PR TITLE
Added support for MQTT protocol

### PR DESCRIPTION
**Pull request by shramos**
Adding support for MQTT (Message Queue Telemetry Transport) protocol by uploading a new protocol layer to scapy/contrib folder.

### CODE REVIEW

**Comment by gpotter2** [comment]
@shramos Hello ! Thanks for your PR.
Could you add tests to your new implementation ? Those are the `.uts` files hanging around.

**Comment by shramos** [comment]
@gpotter2 Hello!, There are some tests.

File: scapy/contrib/mqtt.py [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+                return s[offset + 1:], value
+            if offset > 2:
+                raise Scapy_Exception("%s: malformed length field" %
+                                      self.__class__.__name__)
+
+
+
+# LAYERS
+            
+control_packet_types = {1:'CONNECT (1)',
```
Comment (p-l-) [inline scapy/contrib/mqtt.py]: Do not add the value in the names: `1: 'CONNECT'` is enough. This also applies to other `{value: name}` dictionaries in this code.

File: scapy/contrib/mqtt.py [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+
+
+from scapy.packet import *
+from scapy.fields import *
+from scapy.layers.inet import TCP
+
+
+# CUSTOM FIELDS
+
+class VariableFieldLenField(FieldLenField):
```
Comment (p-l-) [inline scapy/contrib/mqtt.py]: When you add code from e.g. stackoverflow, please be kind enough to say so and add a comment with a link to this code (https://stackoverflow.com/a/43722441/3223422 here).

File: scapy/contrib/mqtt.py [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+
+## LAYERS BINDINGS
+
+bind_layers(TCP,MQTT, sport=1883)
+bind_layers(TCP,MQTT, dport=1883)
+bind_layers(MQTT, MQTT_CONNECT, type=1)
+bind_layers(MQTT, MQTT_CONNACK, type=2)
+bind_layers(MQTT, MQTT_PUBLISH, type=3)
+bind_layers(MQTT, MQTT_PUBACK, type =4)
+bind_layers(MQTT, MQTT_PUBREC, type = 5)
```
Comment (p-l-) [inline scapy/contrib/mqtt.py]: Please respect PEP-8 and be consistent.

File: scapy/contrib/mqtt.py [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+    name = "MQTT fixed header"
+    fields_desc = [
+        BitEnumField("type", 1, 4, control_packet_types),
+        BitEnumField("DUP", 0, 1, {0:'Disabled (0)',
+                                   1:'Enabled (1)'}),
+        BitEnumField("QOS", 0, 2, QOS_level),
+        BitEnumField("RETAIN", 0, 1, {0:'Disabled (0)',
+                                      1:'Enabled (1)'}),
+        VariableFieldLenField("len", None, length_of="len",
+                              adjust=lambda pkt, x: len(pkt.payload),),
```
Comment (p-l-) [inline scapy/contrib/mqtt.py]: Same here about stackoverflow. Also, since this is a hack, a comment explaining it would be more than welcome.

**Comment by shramos** [comment]
Hello!, I've implemented all the changes you requested. Thank you for your comments!

File: scapy/contrib/mqtt.py [hunk 5]
```diff
@@ -1,16 +1,19 @@
-## This file is part of Scapy
-## See http://www.secdev.org/projects/scapy for more informations
-## Copyright (C) Santiago Hernandez Ramos <shramos@protonmail.com>
-## This program is published under GPLv2 license
+''' This file is part of Scapy
```
Comment (p-l-) [inline scapy/contrib/mqtt.py]: This is not a module docstring! This is a copyright notice that should be let as a comment.

File: scapy/contrib/mqtt.py [hunk 6]
```diff
@@ -1,16 +1,19 @@
...
+from scapy.packet import Packet, bind_layers
+from scapy.fields import FieldLenField, BitEnumField, StrLenField, ShortField, \
+    ConditionalField, ByteEnumField, ByteField, StrNullField
 from scapy.layers.inet import TCP
+from scapy.error import Scapy_Exception
 
 
 # CUSTOM FIELDS
 
+# source: http://stackoverflow.com/questions/43711208/scapy-adding-a-new-field-with-a-dynamic-length?noredirect=1&lq=1
```
Comment (p-l-) [inline scapy/contrib/mqtt.py]: Please use a **short** link to the **answer**: https://stackoverflow.com/a/43722441/3223422

File: scapy/contrib/mqtt.py [hunk 7]
```diff
@@ -40,119 +44,122 @@ def getfield(self, pkt, s):
...
-                                   1:'Enabled (1)'}),
+        BitEnumField("DUP", 0, 1, {0: 'Disabled',
+                                   1: 'Enabled'}),
         BitEnumField("QOS", 0, 2, QOS_level),
-        BitEnumField("RETAIN", 0, 1, {0:'Disabled (0)',
-                                      1:'Enabled (1)'}),
+        BitEnumField("RETAIN", 0, 1, {0: 'Disabled',
+                                      1: 'Enabled'}),
+        # Since the size of the len field depends on the next layer, we need to "cheat" with the length_of parameter
+        # and use adjust parameter to calculate the value.
```
Comment (p-l-) [inline scapy/contrib/mqtt.py]: Same here, please link to the source: https://stackoverflow.com/a/43722441/3223422

**Comment by shramos** [comment]
Changes applied! :)

**Comment by p-l-** [comment]
According to Pylint, you should change some constant names and rename layers from `MQTT_XXXXX` to `MQTTXxxxx`.

**Comment by p-l-** [comment]
Also, maybe you could fix errors reported by pycodestyle:
```
scapy/contrib/mqtt.py:1:1: E266 too many leading '#' for block comment
scapy/contrib/mqtt.py:2:1: E266 too many leading '#' for block comment
scapy/contrib/mqtt.py:3:1: E266 too many leading '#' for block comment
scapy/contrib/mqtt.py:4:1: E266 too many leading '#' for block comment
scapy/contrib/mqtt.py:8:80: E501 line too long (80 > 79 characters)
scapy/contrib/mqtt.py:80:80: E501 line too long (116 > 79 characters)
scapy/contrib/mqtt.py:162:80: E501 line too long (83 > 79 characters)
```

**Comment by shramos** [comment]
I think it now fully complies with pylint and pycodestyle.

**Comment by p-l-** [comment]
Thanks a lot!

**Comment by shramos** [comment]
Thank you!

**Pull request merged**